qualtran/_infra/controlled.py

```diff
@@ -318,6 +380,43 @@ def on_classical_vals(self, **vals: 'ClassicalValT') -> Dict[str, 'ClassicalValT
 
         return vals
 
+    def add_my_tensors(
+        self,
+        tn: 'qtn.TensorNetwork',
+        tag: Any,
+        *,
+        incoming: Dict[str, 'SoquetT'],
+        outgoing: Dict[str, 'SoquetT'],
+    ):
+        from qualtran._infra.composite_bloq import _flatten_soquet_collection
+        from qualtran.cirq_interop._cirq_to_bloq import (
+            _tensor_data_from_unitary_and_signature,
+            _tensor_shape_from_signature,
+        )
+
+        # Create an identity tensor corresponding to the signature of current Bloq
+        data = _tensor_data_from_unitary_and_signature(
+            np.eye(2 ** self.signature.n_qubits(), dtype=np.complex128), self.signature
+        )
+        # Verify it has the right shape
+        in_ind = _flatten_soquet_collection(incoming[reg.name] for reg in self.signature.lefts())
+        out_ind = _flatten_soquet_collection(outgoing[reg.name] for reg in self.signature.rights())
+        assert data.shape == tuple(2**soq.reg.bitsize for ind in [out_ind, in_ind] for soq in ind)
+        # Figure out the ctrl indexes for which the ctrl is "active"
+        active_idx: List[Union[int, slice]] = [slice(x) for x in data.shape]
+        ctrl_idx = 0
+        for cv, ctrl_reg in zip(self.ctrl_spec.cvs, self.ctrl_regs):
+            for idx in ctrl_reg.all_idxs():
+                active_idx[ctrl_idx] = int(cv[idx])
+                active_idx[ctrl_idx + len(out_ind)] = int(cv[idx])
```

Comment:
what is going on here? what's the ` + len(out_ind)`? should `active_idx` be two dimensional?

Reply:
We typically pass `out_idx + inp_idx` to the `tn.add()` call. The `active_idx`  is a tuple of length `len(out_idx.shape) + len(inp_idx.shape)`; which is equal to `len(data.shape)` and is used to find the subspace of `data` (an n-d array) that should be actived. 

Thinking of CSWAP;  if you want to express `data[1, :, :, 1, :, :] = _swap_matrix()`; `active_idx` will be a tuple corresponding to `(1, :, :, 1, :, :)`